ocaml/parsing/pprintast.ml

```diff
@@ -332,58 +325,74 @@ let optional_legacy_modalities f m =
     pp_print_space f ()
 
 let mode f m =
-  let {txt; _} = (m : Jane_syntax.Mode_expr.Const.t :> _ Location.loc) in
+  let {txt = Mode txt; _} = m in
   pp_print_string f txt
 
 let modes f m =
-  pp_print_list ~pp_sep:(fun f () -> pp f " ") mode f m.txt
+  pp_print_list ~pp_sep:(fun f () -> pp f " ") mode f m
 
-let maybe_modes_of_type c =
-  let m, cattrs = Jane_syntax.Mode_expr.maybe_of_attrs c.ptyp_attributes in
-  m, { c with ptyp_attributes = cattrs }
+let optional_at_modes f m =
+  match m with
+  | [] -> ()
+  | m -> pp f " %@ %a" modes m
 
-let maybe_modes_type pty ctxt f c =
-  let m, c = maybe_modes_of_type c in
+let optional_atat_modes f m =
   match m with
-  | Some m -> pp f "%a %a" legacy_modes m (pty ctxt) c
-  | None -> pty ctxt f c
+  | [] -> ()
+  | m -> pp f " %@%@ %a" modes m
+
+let maybe_type_at_modes pty ctxt f (c, m) =
+  match m with
+  | _ :: _ -> pp f "%a@ %@@ %a" (pty ctxt) c modes m
+  | [] -> pty ctxt f c
```

Comment:
Since we have now separated `c` and `m` (whereas in the old encoding `m` is encoded in `c`), it's clearer if we print them separately. That is, at each callsite of `maybe_type_at_modes`, we call `core_type c` and `optional_at_modes m` separately.

Similar for `maybe_type_atat_modes` and `maybe_type_atat_modalities`.